pkg/experiment/metastore/compaction_raft_handler.go

```diff
@@ -1,706 +1,197 @@
+			// reassigned, or the request is not valid. This may happen
+			// from time to time, and we should just ignore such requests.
+			revoked++
+		}
+	}
+
+	// AssignJobsMax tells us how many free slots the worker has. We need to
+	// account for the revoked jobs, as they are freeing the worker slots.
+	capacity := int(req.AssignJobsMax) + revoked
+
+	// Next, we need to create new jobs and assign existing
+	//
+	// NOTE(kolesnikovae): On one hand, if we assign first, we may violate the
+	// SJF principle. If we plan new jobs first, it may cause starvation of
+	// lower-priority jobs, when the compaction worker does not keep up with
+	// the high-priority job influx. As of now, we assign jobs before creating
+	// ones. If we change it, we need to make sure that the Schedule
+	// implementation allows doing this.
+	for assigned := 0; assigned < capacity; assigned++ {
+		job, err := scheduler.AssignJob()
 		if err != nil {
-			return err
-		}
-	}
-	for key, levels := range sTable.updatedBlockQueues {
-		for _, l := range levels {
-			queue := h.getOrCreateCompactionBlockQueue(key).blocksByLevel[l]
-			if queue == nil {
-				level.Error(h.logger).Log(
-					"msg", "block queue not found",
-					"shard", key.shard,
-					"tenant", key.tenant,
-					"level", l,
-				)
-				continue
-			}
-			err := h.persistCompactionJobBlockQueue(key.shard, key.tenant, l, queue, tx)
-			if err != nil {
-				return err
-			}
+			level.Error(h.logger).Log("msg", "failed to assign compaction job", "err", err)
+			return nil, err
 		}
-	}
-	for key, jobNames := range sTable.deletedJobs {
-		for _, jobName := range jobNames {
-			jBucket, jKey := keyForCompactionJob(key.shard, key.tenant, jobName)
-			err := updateCompactionJobBucket(tx, jBucket, func(bucket *bbolt.Bucket) error {
-				level.Debug(h.logger).Log(
-					"msg", "deleting job from storage",
-					"job", jobName,
-					"shard", key.shard,
-					"tenant", key.tenant,
-					"storage_bucket", string(jBucket),
-					"storage_key", string(jKey))
-				return bucket.Delete(jKey)
-			})
-			if err != nil {
-				return err
-			}
-		}
-	}
-	for _, jobName := range sTable.updatedJobs {
-		job := h.findJob(jobName)
 		if job == nil {
-			level.Error(h.logger).Log(
-				"msg", "an updated job could not be found",
-				"job", jobName,
-			)
-			continue
-		}
-		err := h.persistCompactionJob(job.Shard, job.TenantId, job, tx)
-		if err != nil {
-			return err
+			// No more jobs to assign.
+			break
 		}
+		p.AssignedJobs = append(p.AssignedJobs, job)
 	}
-	return nil
-}
 
-func (h *CompactionCommandHandler) deleteBlock(tx *bbolt.Tx, shardId uint32, tenant, blockId string) error {
-	for _, meta := range h.index.FindPartitionMetas(blockId) {
-		err := index.UpdateBlockMetadataBucket(tx, meta.Key, shardId, tenant, func(bucket *bbolt.Bucket) error {
-			return bucket.Delete([]byte(blockId))
-		})
+	for created := 0; created < capacity; created++ {
+		plan, err := planner.CreateJob()
```

Comment:
There are a number of reasons:
 - We do want to have small raft messages (where the planned job will end up).
 - We don't want to create jobs ahead of time. This allows altering the planner, scheduler, and workers' configs at any time, with almost instant effect.

The rate I want our system to maintain is at least 1GB/s, which is roughly 256 segments (500ms) per second, or around 1M metadata entries and 50K compaction jobs per hour.

Consider the case when no workers are available (e.g., due to infrastructure issues, misconfiguration, or bugs), or when workers do not have enough capacity to handle all the jobs.

Consider also the case when the metastore is unavailable for a period (e.g., due to infrastructure issues, misconfiguration, or bugs), and our DLQ accumulates a substantial number of entries to process (e.g., 1M, which is a reasonable number). While we will pace the processing, many new blocks (and thus jobs) are to be added over a short period of time.

If there is no limit on the job queue size (which is TODO, BTW), and no limit on how many jobs we can create at once, ingestion could be blocked for a long period of time when the metastore or workers come back online (or added). However, if we produce no more jobs than we can actually handle, this will not happen: the service will adapt to the capacity of the worker fleet, and all jobs will eventually be created and scheduled.

In addition, the underlying implementation of the block queue can handle millions of entries without major issues. While the job queue is simpler, it may cause performance issues if not handled carefully – something we may want to address in the future.

Reply:
I was referring to the choice of `capacity` as the limit, not saying that we should not have a limit. We already assign jobs above so the `capacity` might be "spent" already.

Reply:
The answer is here:

> However, if we produce no more jobs than we *can actually handle*, this will not happen: the service will adapt to the capacity of the worker fleet, and all jobs will eventually be created and scheduled.

I think I need to expand on this and add it to the documentation. The challenge is that we don't know the capacity of our worker fleet in advance, and we have no control over them – they can appear and disappear at any time. Therefore, we need an adaptive approach if we want to keep our queue short and workers busy.

I considered a few options:

1. We produce the `assigned` number of jobs: no workers are assigned any jobs at all. There might be jobs ready for scheduling, but we do not add them to the scheduler because workers are left without assignments due to no jobs being in the scheduler queue. Loop.

2. We produce the `capacity-assigned` number of jobs: we never utilize capacity fully. I'd call this strategy "greedy worker" – the intuition here is that everyone only cares about itself, but in the end, everyone starves.

3. We produce the `capacity` number of jobs: we create jobs for another compaction worker instance. Essentially, we have an evidence that *this* number of jobs will eventually be handled.

4. We ensure that our queue has at least `max_worker_capacity` number of *assignable* jobs in the queue. That works but it is trickier to implement and requires a parameter, which can be dynamic.

There's an unaddressed risk however: if all the workers just abandon all the jobs they take, it will inflate the queue. This is the reason why we need a hard cap and a displacement policy, regardless of the scheduling principles.